sled-agent/src/bootstrap/server.rs

```diff
@@ -142,10 +171,58 @@ impl Inner {
             let log = self.log.new(o!("remote_addr" => remote_addr));
             info!(log, "Accepted connection");
 
+            let sp = self.sp.clone();
+            let trust_quorum = Arc::clone(&trust_quorum);
+            tokio::spawn(async move {
+                match serve_request_after_quorum_initialization(
+                    stream,
+                    sp,
+                    trust_quorum.as_ref(),
+                    &log,
+                )
+                .await
+                {
+                    Ok(()) => info!(log, "Connection closed"),
+                    Err(err) => warn!(log, "Connection failed"; "err" => err),
+                }
+            });
+        }
+    }
+
+    async fn wait_for_sled_initialization(
+        &self,
+    ) -> Result<Arc<Option<ShareDistribution>>, String> {
+        let (tx_share, mut rx_share) = mpsc::channel(1);
+
+        loop {
+            // Wait for either a new client or a response on our channel sent by
+            // a task spawned for a previously-accepted client that provides us
+            // our quorum share.
+            let (stream, remote_addr) = tokio::select! {
```

Comment:
I find the possibility of waiting for a share or a second RSS connection confusing here, especially when considering that the `serve_request_before_quorum_initialization` is forwarding the RSS request to the SledAgent server. If the initial request to the sled agent hasn't completed or timed out, do we really want to allow another concurrent RSS request? I assume the answer is yes, because we want to be able to return a result to the RSS for this new request. 

While it all appears correct, it was hard for me to wrap my head around. I'm not really sure what to do about this to make it easier though.

Reply:
Good question. I considered _not_ spawning `server_request_before_quorum_init()` onto a new task and just serving it inline, which would block any new connection attempts from RSS. I'm still not sure which is right, but like you I lean toward allowing concurrent connections; `BootstrapAgent` internally holds a mutex around the sled agent, which will prevent concurrent RSS requests from actually doing concurrent (conflicting) operations. If we get multiple RSS requests simultaneously, that mutex will serialize them, and the first one that succeeds will win. Any subsequent requests that match the winner will also succeed (with no work done), and any that don't match will fail.

Reply:
This all makes sense. Thank you for the extra explanation.